crates/matrix-sdk/src/room/mod.rs

```diff
@@ -375,61 +381,50 @@ impl Room {
-            });
+        while let Some(predecessor) = current_room.predecessor_room() {
+            let maybe_predecessor_room = current_room.client.get_room(&predecessor.room_id);
 
-            match pred {
-                None => break 'leave_aggregate_loop,
-                Some(room) => match room {
-                    Ok(room) => {
-                        rooms.push(room);
-                        cur_room = rooms.last().expect("rooms just pushed so can't be empty");
-                    }
-                    Err(e) => {
-                        debug!("Tombstone following terminated early: {e:?}");
-                        break 'leave_aggregate_loop;
-                    }
-                },
+            if let Some(predecessor_room) = maybe_predecessor_room {
+                rooms.push(predecessor_room.clone());
+                current_room = rooms.last().expect("Room just pushed so can't be empty");
+            } else {
+                warn!("Cannot find predecessor room");
+                break;
             }
         }
 
-        let mut rooms_iter = rooms.into_iter();
         let batch_size = 5;
 
-        'leave_action_loop: loop {
-            let batch: Vec<Room> = rooms_iter.by_ref().take(batch_size).collect();
-            if batch.is_empty() {
-                break 'leave_action_loop;
-            }
+        let rooms_futures: Vec<_> = rooms
+            .iter()
+            .filter_map(|room| match room.state() {
+                RoomState::Joined | RoomState::Invited | RoomState::Knocked => {
+                    Some(room.leave_impl())
+                }
+                RoomState::Banned | RoomState::Left => None,
+            })
+            .collect();
 
-            let batch = batch
-                .iter()
-                .map(|room| match room.state() {
-                    RoomState::Joined => Ok(Some(room.leave_this())),
-                    RoomState::Invited => Ok(Some(room.leave_this())),
-                    RoomState::Knocked => Ok(Some(room.leave_this())),
-                    RoomState::Left => Ok(None),
-                    _ => Err(Error::PredecessorWrongState),
-                })
-                .collect::<Result<Vec<Option<_>>, _>>()?
-                .into_iter()
-                .flatten()
-                .collect::<Vec<_>>();
-
-            if join_all(batch).await.iter().any(Result::is_err) {
-                error!("Error occurred while leaving room.");
-                return Err(Error::ErrorDuringRoomLeave);
+        let mut futures_stream = futures_stream::iter(rooms_futures).buffer_unordered(batch_size);
+
+        let mut maybe_this_room_failed_with: Option<Error> = None;
+
+        while let Some(result) = futures_stream.next().await {
+            if let (Err(e), room) = result {
+                if room.room_id() == self.room_id() {
+                    maybe_this_room_failed_with = Some(e);
+                    continue;
+                }
+                warn!("{e:?}");
```

Comment:
We should explain what happened here, instead of just showing the error. The error will just show that a request failed with a specific reason, but we won't know why we made that request. 

This could also be put in an else branch which makes the `continue` in the if branch unnecessary.